### PR TITLE
Feature/filter component

### PR DESCRIPTION
**Pull request by jeffsupancic**
## Related Issue
#1311 #379 #333

## Description
Implement feature request that I submitted, "Filter by a range of data" #1311 by allowing the developer to pass a filter component which could be a range filter.

## Impacted Areas in Application
Update to m-table-filter-row.js and src/prop-types.js to handle rendering and added very basic example to the demo.js

### CODE REVIEW

**Comment by sureshtidbit** [comment]
hey @jeffsupancic @mbrn can you please check if its work then i think many filter problem will be solved

**Comment by Cristy94** [comment]
Is this functional/usable? Any progress on the merge?

**Comment by jeffsupancic** [comment]
@Cristy94 Yes it works, we're actually using it professionally in the fork for now until the merge is done.  No idea on when it will be merged though.

@mbrn Any idea on when we can get this PR reviewed?

File: demo/components/filter-component.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
```
Comment (mbrn) [inline demo/components/filter-component.js]: I won't add this file to pr. The idea is let developers to override filterComponent.

**Comment by nicolasc38** [comment]
Hello guys! @jeffsupancic @mbrn 
First of all, thanks for all the great work on material-table.

I am currently working on a project where i would like to filter a column by a date range.
I thought i would use your new filter component feature but I don't get how to implement that.

Indeed in your demo/components/filter-component.js file, you use the method :
` onChange={(event) => {
        onFilterChanged(columnDef.tableData.id, event.target.value);
      }}`

My guess was that to work with date range you have to override this behavior not to work with one value (event.target.value) but with a range of values?
Could you tell me if that's the right way to do so and if yes maybe how to get started please?

Ps : this is my first github post, hope I wrote it on the proper section and my question isn't trivial :)

**Comment by ctbucha** [comment]
FYI for other people looking to have a range filter, I implemented a range filter for simple numeric fields. It's not perfect since it uses lodash and doesn't support localization, but it works well enough right now for my use case. Building a range filter for dates should be fairly similar.

```
import React from 'react';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import InputAdornment from '@material-ui/core/InputAdornment';
import FilterIcon from '@material-ui/icons/FilterList';
import Tooltip from '@material-ui/core/Tooltip';
import * as _ from 'lodash';

export default function FilterNumericBetween({ columnDef, onFilterChanged }) {
  return (
    <>
    <TextField
      style={columnDef.type === 'numeric' ? { float: 'right' } : {}}
      type={columnDef.type === 'numeric' ? 'number' : 'search'}
      value={_.get(columnDef, ['tableData', 'filterValue', 'greaterThan']) || ''}
      placeholder={columnDef.filterPlaceholder || ''}
      onChange={(event) => {
        const value = {...columnDef.tableData.filterValue};
        value.greaterThan = event.target.value;
        onFilterChanged(columnDef.tableData.id, value);
      }}
      InputProps={columnDef.hideFilterIcon ? undefined : {
        startAdornment: (
          <InputAdornment position="start">
            <Tooltip title="Filter greater than">
              <div style={{display: 'flex'}}>
                <FilterIcon />
                <Typography>{'>'}</Typography>
              </div>
            </Tooltip>
          </InputAdornment>
        )
      }}
    />
    <TextField
      style={columnDef.type === 'numeric' ? { float: 'right' } : {}}
      type={columnDef.type === 'numeric' ? 'number' : 'search'}
      value={_.get(columnDef, ['tableData', 'filterValue', 'lessThan']) || ''}
      placeholder={columnDef.filterPlaceholder || ''}
      onChange={(event) => {
        const value = {...columnDef.tableData.filterValue};
        value.lessThan = event.target.value;
        onFilterChanged(columnDef.tableData.id, value);
      }}
      InputProps={columnDef.hideFilterIcon ? undefined : {
        startAdornment: (
          <InputAdornment position="start">
            <Tooltip title="Filter less than">
              <div style={{display: 'flex'}}>
                <FilterIcon />
                <Typography>{'<'}</Typography>
              </div>
            </Tooltip>
          </InputAdornment>
        )
      }}
    />
    </>
  );
}
```

```
<MaterialTable
  options={{ filtering: true }}
  columns: [{title: 'My number', field: 'myNumber', type: 'numeric', filterComponent: FilterNumericBetween }]
  data={async(query) => /* actually filter data here */}
/>
```

**Comment by jeffsupancic** [comment]
Thanks @ctbucha that's a good example.

Another thing to note @nicolasc38 you will likely also want to implement a customFilterAndSearch function for the column to handle the range.

**Comment by nicolasc38** [comment]
Thank you guys I'll look into it! :+1: @ctbucha @jeffsupancic

**Comment by mariostasos** [comment]
@ctbucha I had tried your solution and I have a question how are you using these `greaterThan`
`lessThan` ? It seems that these do not exist in of the `m-table-filter-row.js`. I am using version 1.59.0

**Comment by Amber711** [comment]
Hi @ctbucha,  when you call onFilterChanged(columnDef.tableData.id, value) inside the change event of the custom filter, how can we actually catch the filter terms user has input in data={async(query) => /* actually filter data here */  ? because in my case, it's a server side filtering. I need to pass whatever user input to backend and fetch the new data.

Thank you!

**Comment by jeffsupancic** [comment]
@mariostasos 

> @ctbucha I had tried your solution and I have a question how are you using these `greaterThan`
> `lessThan` ? It seems that these do not exist in of the `m-table-filter-row.js`. I am using version 1.59.0

You can pass an object to `onFilterChanged` as the value.  That object will be available in the column's `customFilterAndSearch` function.

**Comment by jeffsupancic** [comment]
> Hi @ctbucha, when you call onFilterChanged(columnDef.tableData.id, value) inside the change event of the custom filter, how can we actually catch the filter terms user has input in data={async(query) => /* actually filter data here */ ? because in my case, it's a server side filtering. I need to pass whatever user input to backend and fetch the new data.
> 
> Thank you!

You'll be able to access the value passed by looking at `query.filters`

**Comment by BananaCHIU** [comment]
@ctbucha It seems that after implementing the range filter, other column filters are not working. I have to implement all other filters in the function async(query)=>{}

**Comment by Dos771** [comment]
Who is looking for date range, you can use that code. 

`import React from 'react';
import TextField from '@material-ui/core/TextField';
import * as _ from 'lodash';

export function isDateInRange(dateFrom, dateTo, dateToBeFiltered) {
  const parsedDateToBeFiltered = parseDateWithDots(dateToBeFiltered);

  if (!parsedDateToBeFiltered) {
    return !(dateFrom || dateTo);
  }

  if (!(dateFrom || dateTo)) {
    return true;
  } else if (!dateFrom) {
    return parsedDateToBeFiltered <= new Date(dateTo).setHours(0, 0, 0, 0);
  } else if (!dateTo) {
    return parsedDateToBeFiltered >= new Date(dateFrom).setHours(0, 0, 0, 0);
  }
  return parsedDateToBeFiltered >= new Date(dateFrom).setHours(0, 0, 0, 0)
    && parsedDateToBeFiltered <= new Date(dateTo).setHours(0, 0, 0, 0);
}

export function parseDateWithDots(dateString) {
  if (!dateString) {
    return null;
  }

  if (dateString.includes(' ')) {
    dateString = dateString.split(' ')[0];
  }

  const dateArray = dateString.split('.');
  return new Date(dateArray[2], dateArray[1] - 1, dateArray[0]);
}



export const FilterNumericBetween = (props) => {
  const { columnDef, onFilterChanged } = props 

  return (
    <>
      <TextField
        type={columnDef.type === 'numeric' ? 'number' : 'search'}
        value={_.get(columnDef, ['tableData', 'filterValue', 'dateFrom']) || ''}
        placeholder={columnDef.filterPlaceholder || ''}
        onChange={(event) => {
          let value = {...columnDef.tableData.filterValue};
          value.dateFrom = event.target.value;
          onFilterChanged(columnDef.tableData.id, value);
        }}
      />
      
      <TextField
        type={columnDef.type === 'numeric' ? 'number' : 'search'}
        value={_.get(columnDef, ['tableData', 'filterValue', 'dateTo']) || ''}
        placeholder={columnDef.filterPlaceholder || ''}
        onChange={(event) => {
          let value = {...columnDef.tableData.filterValue};
          value.dateTo = event.target.value;
          onFilterChanged(columnDef.tableData.id, value);
        }}
      />
    </>
  );
}

const [column, setColumn] = useState([
  {title: 'Дата', field: 'startDate',  filterComponent: FilterNumericBetween, customFilterAndSearch: (filter, rowData) => isDateInRange(filter.dateFrom, filter.dateTo, rowData.startDate)},
])
`

**Pull request merged**